src/DiskIO/IpcIo/IpcIoFile.cc

```diff
@@ -538,18 +569,13 @@ IpcIoFile::StatQueue(std::ostream &os)
 void
 IpcIoFile::OpenTimeout(void *const param)
 {
-    Must(param);
-    // the pointer is used for comparison only and not dereferenced
-    const IpcIoFile *const ipcIoFile =
-        reinterpret_cast<const IpcIoFile *>(param);
-    for (IpcIoFileList::iterator i = WaitingForOpen.begin();
-            i != WaitingForOpen.end(); ++i) {
-        if (*i == ipcIoFile) {
-            (*i)->openCompleted(NULL);
-            WaitingForOpen.erase(i);
-            break;
-        }
+    while (!WaitingForOpen.empty() && current_time >= WaitingForOpen.begin()->first) {
+        const auto file = WaitingForOpen.begin()->second;
+        file->openCompleted(nullptr);
+        WaitingForOpen.erase(WaitingForOpen.begin());
```

Comment:
I think we ought to reverse the order here because some future openCompleted() might schedule some more timeouts, affecting the value of begin(). That is unlikely but theoretically possible and there is no reason to expose ourselves to those dangers.

```suggestion
        WaitingForOpen.erase(WaitingForOpen.begin());
        file->openCompleted(nullptr);
```